File: app/src/main/java/com/gystry/pjetpack/ui/AbsListFragment.java
package com.gystry.pjetpack.ui;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProviders;
import androidx.paging.PagedList;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.gystry.libcommon.EmptyView;
import com.gystry.pjetpack.AbsViewModel;
import com.gystry.pjetpack.R;
import com.gystry.pjetpack.databinding.LayoutRefreshViewBinding;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.scwang.smartrefresh.layout.api.RefreshLayout;
import com.scwang.smartrefresh.layout.constant.RefreshState;
import com.scwang.smartrefresh.layout.listener.OnLoadMoreListener;
import com.scwang.smartrefresh.layout.listener.OnRefreshListener;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * @author gystry
 * 创建日期：2020/7/9 09
 * 邮箱：gystry@163.com
 * 描述：
 */
public abstract class AbsListFragment<T, M extends AbsViewModel<T>> extends Fragment implements OnRefreshListener, OnLoadMoreListener {
    private LayoutRefreshViewBinding binding;
    public EmptyView emptyView;
    protected RecyclerView recyclerView;
    protected SmartRefreshLayout refreshLayout;
    protected PagedListAdapter<T, RecyclerView.ViewHolder> adapter;
    protected M mViewModel;
    private DividerItemDecoration decoration;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = LayoutRefreshViewBinding.inflate(inflater, container, false);
        emptyView = binding.emptyView;
        recyclerView = binding.recyclerView;
        refreshLayout = binding.refreshLayout;

        refreshLayout.setEnableLoadMore(true);
        refreshLayout.setEnableRefresh(true);
        refreshLayout.setOnRefreshListener(this);
        refreshLayout.setOnLoadMoreListener(this);

        adapter = getAdapter();
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        recyclerView.setItemAnimator(null);
        //默认给列表中的Item 一个 10dp的ItemDecoration
        decoration = new DividerItemDecoration(getContext(), LinearLayoutManager.VERTICAL);
        decoration.setDrawable(ContextCompat.getDrawable(getContext(), R.drawable.list_divider));
        recyclerView.addItemDecoration(decoration);

        genericViewModel();
        return binding.getRoot();
    }

    private void genericViewModel() {
        //利用 子类传递的 泛型参数实例化出absViewModel 对象。
        ParameterizedType type = (ParameterizedType) getClass().getGenericSuperclass();
        Type[] arguments = type.getActualTypeArguments();
        if (arguments.length > 1) {
            Type argument = arguments[1];
            Class modelClaz = ((Class) argument).asSubclass(AbsViewModel.class);
            mViewModel = (M) ViewModelProviders.of(this).get(modelClaz);

            //触发页面初始化数据加载的逻辑
            mViewModel.getPageData().observe(this, pagedList -> submitList(pagedList));

            //监听分页时有无更多数据,以决定是否关闭上拉加载的动画
            mViewModel.getBoundaryPageData().observe(this, hasData -> finishRefresh(hasData));
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
//        ParameterizedType type = (ParameterizedType) getClass().getGenericSuperclass();
//        Type[] typeArguments = type.getActualTypeArguments();
//        if (typeArguments.length > 1) {
//            Type argument = typeArguments[1];
//            Class viewModelClaz = ((Class) argument).asSubclass(AbsViewModel.class);
//            mViewModel = (M) ViewModelProviders.of(this).get(viewModelClaz);
//            mViewModel.getPageData().observe(getViewLifecycleOwner(), new Observer<PagedList<T>>() {
//                @Override
//                public void onChanged(PagedList<T> pagedList) {
//                    adapter.submitList(pagedList);
//                }
//            });
//            mViewModel.getBoundaryPageData().observe(getViewLifecycleOwner(), new Observer<Boolean>() {
//                @Override
//                public void onChanged(Boolean aBoolean) {
//                    finishRefresh(aBoolean);
//                }
//            });
//        }
    }

    public void submitList(PagedList<T> pagedList) {
        if (pagedList.size() > 0) {
            adapter.submitList(pagedList);
        }
        finishRefresh(pagedList.size() > 0);
    }

    public void finishRefresh(boolean hasData) {
        PagedList<T> currentList = adapter.getCurrentList();
        hasData = hasData || currentList != null && currentList.size() > 0;
        RefreshState state = refreshLayout.getState();
        if (state.isFooter && state.isOpening) {
            refreshLayout.finishLoadMore();
        } else if (state.isHeader && state.isOpening) {
            refreshLayout.finishRefresh();
        }

        if (hasData) {
            emptyView.setVisibility(View.GONE);
        } else {
            emptyView.setVisibility(View.VISIBLE);
        }
    }

    public abstract PagedListAdapter<T, RecyclerView.ViewHolder> getAdapter();

}

File: appkt/src/main/java/com/gystry/appkt/ui/AbsListFragment.kt
package com.gystry.appkt.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProviders
import androidx.paging.PagedList
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.gystry.appkt.AbsViewModel
import com.gystry.appkt.R
import com.gystry.appkt.databinding.LayoutRefreshViewBinding
import com.gystry.libcommon.EmptyView
import com.scwang.smartrefresh.layout.SmartRefreshLayout
import com.scwang.smartrefresh.layout.api.RefreshLayout
import com.scwang.smartrefresh.layout.listener.OnLoadMoreListener
import com.scwang.smartrefresh.layout.listener.OnRefreshListener
import java.lang.reflect.ParameterizedType

/**
 * @author gystry
 * 创建日期：2021/3/31 10
 * 邮箱：gystry@163.com
 * 描述：
 */
abstract class AbsListFragment<T, M : AbsViewModel<T>,V:RecyclerView.ViewHolder> : Fragment(), OnLoadMoreListener, OnRefreshListener {
    private lateinit var adapter: PagedListAdapter<T, V>
    protected lateinit var binding: LayoutRefreshViewBinding
    protected lateinit var recyclerView: RecyclerView
    protected lateinit var refreshLayout: SmartRefreshLayout
    protected lateinit var emptyView: EmptyView
    protected lateinit var mViewModel:M

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = LayoutRefreshViewBinding.inflate(inflater, container, false)
        recyclerView = binding.recyclerView
        refreshLayout = binding.refreshLayout
        emptyView = binding.emptyView

        refreshLayout.setEnableLoadMore(true)
        refreshLayout.setEnableRefresh(true)
        refreshLayout.setOnLoadMoreListener(this)
        refreshLayout.setOnRefreshListener(this)

        recyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        recyclerView.itemAnimator = null
        adapter = getAdapter()
        //默认给列表中的Item 一个 10dp的ItemDecoration
       val decoration = DividerItemDecoration(context, LinearLayoutManager.VERTICAL)
        decoration.setDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.list_divider)!!)
        recyclerView.addItemDecoration(decoration)
        recyclerView.adapter = adapter

        afterCreateView()
        return binding.root
    }

    abstract fun afterCreateView()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val type = javaClass.genericSuperclass as ParameterizedType
        val arguments = type.actualTypeArguments
        if(arguments.size>1){
            val argument = arguments[1]
            val modelClaz = (argument as Class<*>).asSubclass(AbsViewModel::class.java)
            mViewModel = ViewModelProviders.of(this).get(modelClaz) as M
            mViewModel.pageData.observe(viewLifecycleOwner){
                adapter.submitList(it)
            }
            mViewModel.boundaryPageData.observe(viewLifecycleOwner){
                finishRefresh(it)
            }
        }
    }

    open fun submitList(pageList: PagedList<T>) {
        if (pageList.size > 0) {
            adapter.submitList(pageList)
        }
        finishRefresh(pageList.size > 0)
    }

    open fun finishRefresh(hasData: Boolean) {
        var hasData = hasData
        val currentList = adapter.currentList
        hasData = hasData || (currentList != null && currentList.size > 0)
        var state = refreshLayout.state
        if (state.isFooter && state.isOpening) {
            refreshLayout.finishLoadMore()
        } else if (state.isHeader && state.isOpening) {
            refreshLayout.finishRefresh()
        }

        if (hasData) {
            emptyView.visibility = View.GONE
        } else {
            emptyView.visibility = View.VISIBLE
        }
    }

   abstract fun getAdapter(): PagedListAdapter<T, V>

    override fun onLoadMore(refreshLayout: RefreshLayout) {
    }

    override fun onRefresh(refreshLayout: RefreshLayout) {
    }
}